src/librustc/ty/maps.rs

```diff
@@ -463,13 +479,12 @@ macro_rules! define_maps {
         }
 
         impl<$tcx> Maps<$tcx> {
-            pub fn new(dep_graph: DepGraph,
-                       providers: IndexVec<CrateNum, Providers<$tcx>>)
+            pub fn new(providers: IndexVec<CrateNum, Providers<$tcx>>)
                        -> Self {
                 Maps {
                     providers,
                     query_stack: RefCell::new(vec![]),
-                    $($name: RefCell::new(DepTrackingMap::new(dep_graph.clone()))),*
+                    $($name: RefCell::new(QueryMap::new())),*
```

Comment:
I'm wondering if there's a way to get rid of the `RefCell`'s here - perhaps a type with a very simple foolproof API based on `Cell` - something that requires a trait implemented for `Copy` types and `Rc` (which can't run arbitrary code when cloning), perhaps? It would probably have to be in its own module to prevent accidental bypass.